### ISSUE TITLE
RAMP adding/removing scripts to the html document

### ISSUE THREAD

**Issue by teaganryley** [fgpv-vpgf/fgpv-vpgf#4091]
Hello! I've noticed that RAMP adds a large number of scripts to the head of the html document on load (see attached):

![ramp-scripts](https://user-images.githubusercontent.com/29206084/214174287-c2c89d5b-df65-4e6b-bfbc-e266e097764c.png)

When the user navigates to another route (which does not contain a RAMP instance) these scripts are still attached to the document. On the other hand, refreshing the browser will remove the scripts. 

Is this expected behaviour? If yes, what is the reason for this behaviour? Finally, under what circumstances will RAMP remove its dependencies from the document?

**Comment by prosperva** [comment]
This could be a behaviour with React or Angular since you might be looking at a single page application.

**Comment by teaganryley** [comment]
>If you are using React to load RAMP scripts, you will want to load them when the component mounts and then remove them from DOM when the component unmounts.

Right, so I was using React to mount/unmount `rv-main.js`, `jquery`, `rv-styles.css`, and the polyfill. When testing that code, I realized that RAMP was adding its own scripts to the document (the ~20 or so scripts in the screenshot above).

The brute-force solution would be: on component unmount, iterate through the scripts attached to the DOM. Remove the dependencies above, as well as any scripts from the `js.arcgis.com` domain. However, I'm open to other suggestions.

**Comment by james-rae** [comment]
I can provide a bit of background on what's likely happening here.

Those scripts are being added by the ESRI JS API. RAMP [injects the API loader](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/master/packages/ramp-geoapi/src/index.js#L140) (e.g. `https://js.arcgis.com/3.41/init.js`) into the `head` tag. The scripts you screenshotted are likely caused by the logic in `init.js` combined with the calls to Dojo's `request` for each ESRI module  (see [here](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/master/packages/ramp-geoapi/src/index.js#L20)).

RAMP loads the API in this manner to avoid conflicts between Dojo and Angular. ESRI API v3 requires Dojo, and back when this was being built it would cause mayhem at start time. To solve it, we had the `geoapi` be a stand-alone module, and had the loading of ESRI/Dojo be part of the module's startup. RAMP would then innocently load the module and not notice Dojo was hiding inside it.

At the time, no consideration was made for routing or anything like that, we were just building typical pages that hosted a map.

So fun stuff, your site likely ends up with Dojo wrapped in Angular wrapped in React. 

Other than your brute force solution, the only other thing I could recommend is to fork the codebase, and modify the section of code I linked above. You could possibly add some global page flag to look for that would skip the "load the ESRI API" steps; your React stuff would set the flag as you deem fit.

**Comment by prosperva** [comment]
@james-rae always a pleasure reading your insights into RAMP. I think RAMP team moved from Angular to views, correct me if I am wrong.

My one million dollar question is this: how much work would it take to convert the Angular code or vuejs to a package or is it possible to use the angular or vue is in a typescript based project?

**Comment by james-rae** [comment]
Correct, [RAMP4](https://github.com/ramp4-pcar4/ramp4-pcar4) has migrated to Vue v3. It also was able to drop Dojo by updating to ESRI v4.

We have an open R&D item to look into making RAMP4 a package ( https://github.com/ramp4-pcar4/ramp4-pcar4/discussions/1495 ); however at the moment most of our resources are being pulled elsewhere. Our resident smart-guy for package type stuff will hopefully be re-allocated to our team after April. 

I also _BELIEVE_ that RAMP4 provides a build that is more package-like, you can import from it in typescript. See [this doc page](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/main/docs/api/startup.md#ramp-build-files), and note there is a type-o that I'll correct on my next commit (`ramp.global.js` has been renamed to just `ramp.js`). I'm not sure if typescript definitions are exported via the `ramp.esm.js` file or if that's an additional thing for us to explore. However, at this time, that `esm.js` cannot be just added to a `package.json` sort of file (or at least without some trickery).
